source/FormationPattern.cpp

```diff
@@ -93,7 +205,228 @@ void FormationPattern::SetName(const std::string &name)
+Point FormationPattern::Position(unsigned int ring, unsigned int lineNr, unsigned int repeatNr,
+	unsigned int linePosition, double diameterToPx, double widthToPx, double heightToPx) const
+{
+	// First check if the inputs result in a valid line or arc position.
+	if(lineNr >= lines.size())
+		return Point();
+	const Line &line = lines[lineNr];
+	if(ring > 0 && repeatNr >= line.repeats.size())
+		return Point();
+
+	// Perform common start and end/anchor position calculations in pixels.
+	Point startPx = line.start.GetPx(diameterToPx, widthToPx, heightToPx);
+	Point endOrAnchorPx = line.endOrAnchor.GetPx(diameterToPx, widthToPx, heightToPx);
+
+	// Get the number of positions for this line or arc.
+	int positions = line.positions;
+
+	// Check if we have a valid repeat section and apply it to the common calculations if we have it.
+	const LineRepeat *repeat = nullptr;
+	if(ring > 0 && repeatNr < line.repeats.size())
+	{
+		repeat = &(line.repeats[repeatNr]);
+		startPx += repeat->repeatStart.GetPx(diameterToPx, widthToPx, heightToPx) * ring;
+		endOrAnchorPx += repeat->repeatEndOrAnchor.GetPx(diameterToPx, widthToPx, heightToPx) * ring;
+		positions += repeat->repeatPositions * ring;
+	}
+
+	// Compensate for any skipped positions. This would usually be the start-position on the line, but it can be the
+	// end position if we are on an alternating line.
+	if(ring % 2 && repeat && repeat->alternating && line.skipLast)
+		++linePosition;
+	else if(line.skipFirst)
+		++linePosition;
+
+	// Switch to arc-specific calculations if this line is an arc.
+	if(line.isArc)
+	{
+		// Calculate angles and radius from anchor to start.
+		double startAngle = Angle(startPx).Degrees();
+		double endAngle = line.angle;
+		double radius = startPx.Length();
+
+		// Apply repeat section for endAngle, startAngle and anchor were already done before.
+		if(repeat)
+		{
+			endAngle += repeat->repeatAngle * ring;
+
+			// Calculate new start and turn end angle if we need to alternate this repeat.
+			if(ring % 2 && repeat->alternating && positions > 0)
+			{
+				startAngle += endAngle;
+				endAngle = -endAngle;
+			}
+		}
+
+		// Apply positions to get the correct position-angle.
+		if(positions > 1)
+			endAngle /= positions - 1;
+		double positionAngle = startAngle + endAngle * linePosition;
+
+		// Get into the range of 0 to 360 for conversion to angle.
+		if(positionAngle < 0)
+			positionAngle = -fmod(-positionAngle, 360) + 360;
+		else
+			positionAngle = fmod(positionAngle, 360);
+
+		// Combine anchor with the position and return the result.
+		return endOrAnchorPx + Angle(positionAngle).Unit() * radius;
+	}
+
+	// This is not an arc, perform the line-based calculations.
+
+	// Swap start and end if we need to alternate in the repeat section.
+	// Apply repeat section if it is relevant.
+	if(ring % 2 && repeat && repeat->alternating)
+	{
+		Point tmpPx = endOrAnchorPx;
+		endOrAnchorPx = startPx;
+		startPx = tmpPx;
+	}
```

Comment:
How about a std::swap here?
```suggestion
		swap(startPx, endOrAnchorPx);
```